File: src/main/scala/com/github/alikemalocalan/tunnel/HttpProxyRemoteHandler.scala
package com.github.alikemalocalan.tunnel

import io.netty.buffer.Unpooled
import io.netty.channel.{Channel, ChannelFuture, ChannelHandlerContext, ChannelInboundHandlerAdapter}
import org.slf4j.LoggerFactory

class HttpProxyRemoteHandler(clientChannel: Channel) extends ChannelInboundHandlerAdapter {
  val logger = LoggerFactory.getLogger(this.getClass)

  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit =
    clientChannel.writeAndFlush(msg)
      .addListener { future: ChannelFuture =>
        if (future.isSuccess) ctx.channel().read()
        else future.channel().close()
      }

  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
    logger.error("Shit happen at Server Connection", cause)
    ctx.close()
  }

  override def channelActive(ctx: ChannelHandlerContext): Unit = {
    ctx.read();
    ctx.write(Unpooled.EMPTY_BUFFER);
  }
}

File: src/main/kotlin/com/github/alikemalocalan/greentunnel4jvm/HttpProxyRemoteHandler.kt
package com.github.alikemalocalan.greentunnel4jvm

import io.netty.buffer.Unpooled
import io.netty.channel.Channel
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import org.slf4j.Logger
import org.slf4j.LoggerFactory


class HttpProxyRemoteHandler(private val clientChannel: Channel) : ChannelInboundHandlerAdapter() {
    private val logger: Logger = LoggerFactory.getLogger(this::class.java)

    override fun channelInactive(ctx: ChannelHandlerContext?) {
        flushAndClose(clientChannel)
    }

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        clientChannel.writeAndFlush(msg)
    }

    override fun channelActive(ctx: ChannelHandlerContext) {
        ctx.read()
        ctx.write(Unpooled.EMPTY_BUFFER)
    }

    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        flushAndClose(clientChannel)
        logger.error("Website Connection error")
    }

    private fun flushAndClose(ch: Channel?) {
        if (ch != null && ch.isActive) {
            ch.writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
        }
    }


}